### PR TITLE
Check testing framework in action text installer

### LINKED ISSUE #35053: action_text:install task copies default fixture to /test regardless of testing framework
### Steps to reproduce
- Have a Rails 6.0.0.beta1 application with a third-party testing framework (`rspec` in this case)
- Run `rails action_text:install`

### Expected behavior
The task can identify the testing framework used (f.x. `rspec`), and copy the default fixture to the right location (`/spec/fixtures`).
Perhaps adding an optional argument to the task to select a testing framework/location manually could help.

### Actual behavior
The task copies the default fixture to `/test/fixtures`, regardless of the testing framework used.

### System configuration
**Rails version**:
6.0.0.beta1

**Ruby version**:
2.6.0

### LINKED ISSUE #35305: action text  genrate test/fixtures even test skiped in new app
### Steps to reproduce
 New rails app skip mini test with  -T option: `rails new my_app -T`
 add  action text : `rails action_text:install`
### Expected behavior
should  not generate test cases
### Actual behavior
test/fixtures/action_text/rich_texts.yml
is genrated
### System configuration
**Rails version**:
6.0.0 beta1
**Ruby version**:
2.6.1

### PR DESCRIPTION
**Pull request by vinistock**
### Summary

Fixes #35053

The goal of this PR is to check if the "spec" directory has been registered in an application and then properly adapt the action_text installer to use the adequate folder so that fixtures don't end up under "test".

I tested this in a local application and it worked, but please let me know if this is the expected implementation.

### CODE REVIEW

**Comment by y-yagi** [comment]
Since this file is needed at the time of the test run. So in my opinion, it is better to generate it together when the generator generates the test file.
https://github.com/rails/rails/blob/23125378673bcc606b274027666a126573e136f8/actionmailbox/lib/rails/generators/test_unit/mailbox_generator.rb#L11 

This will also allow the library(e.g. `rspec-rails`) to add fixtures at the right time. What do you think?

**Comment by vinistock** [comment]
@y-yagi No problem at all. Do you believe the implementation looks good?

File: actiontext/lib/templates/installer.rb [hunk 1]
```diff
@@ -8,11 +8,17 @@
 JS_DEPENDENCIES = JS_PACKAGE["peerDependencies"].dup.merge \
   JS_PACKAGE["name"] => "^#{JS_PACKAGE["version"]}"
 
+if Rails::SourceAnnotationExtractor::Annotation.directories.include?("spec")
```
Comment (kaspth) [inline actiontext/lib/templates/installer.rb]: This definitely can’t be right. I do think we’d have to look at generators and their hooks to get this right.

Comment (vinistock) [reply actiontext/lib/templates/installer.rb]: @kaspth do you have an example I could learn from? Or can you further explain the kind of implementation you're expecting?

**Comment by rafaelfranca** [comment]
I really thing this is being done at the wrong level. This installer should be a generator, and being a generator we could have different hooks to different test frameworks like others generators have.

**Comment by vinistock** [comment]
@rafaelfranca thanks for the guidance, it helped to get me started.

I've created basic generators using the hooks you mentioned. I have a few doubts before I can continue developing these.

1. With this implementation, the generator runs by running `rails g actiontext`. Is this the desired or should it be `rails g actiontext:install`?
2. Should the javascript portion of the original installer (installing via yarn and appending to the package file) be inside this generator as well? What is the standard way of running a bin file from within a generator?
3. For copying yarn.tt, is there any way I can reuse the app generator? At the moment, I'm simply copying the template, but the app generator runs chmod and other bin related operations that I could reuse.

**Comment by vinistock** [comment]
When fully implemented, I believe this PR will fix #35053, #35003 and #35305.

**Comment by y-yagi** [comment]
It is my personal opinion. 

1. The command the user should execute should remain `action_text:install`. Therefore, I do not think that it is necessary for the name of the generator to be `action_text:install` (I think that it is more important that the process is properly divided for each process). 
2. I think it better to do this. Since the generator is based on thor, you can use the method of thor (in this case surely `run`). 
3. I'm not sure, why need to copy `yarn.tt`? In my feeling that is not should do in the install task.

**Comment by vinistock** [comment]
@y-yagi @rafaelfranca here's the current state of the PR: running `rake action_text:install` will add javascript dependencies, create migrations and templates. I used a local application to try it out and it worked.

Open questions:
1. Should the yarn template be copied in the current way, not be copied or be copied using the app generator?

**Comment by y-yagi** [comment]
My answer remains the same, which is not necessary for this task. I think it is not a process that should be done in this task.

**Comment by vinistock** [comment]
@y-yagi I have pushed a commit to stop copying the yarn template in this generator. Let me know if there are other concerns.

**Comment by vinistock** [comment]
@rafaelfranca do you have any other concerns or requests? I'm available for further improvements so that we can have this merged.

File: actiontext/lib/rails/generators/actiontext/actiontext_generator.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+require "pathname"
+require "json"
+
+module Rails
+  module Generators
+    class ActiontextGenerator < Base
+      JS_PACKAGE_PATH = Pathname.new("#{__dir__}/../../../../package.json")
+
+      JS_PACKAGE = JSON.load(JS_PACKAGE_PATH)
```
Comment (rafaelfranca) [inline actiontext/lib/rails/generators/actiontext/actiontext_generator.rb]: Can we move this to a method? I don't think we should be reading a JSON when this file is loaded.

File: actiontext/lib/tasks/actiontext.rake [hunk 3]
```diff
@@ -8,8 +8,8 @@ namespace :action_text do
   task install: %w( environment run_installer copy_migrations )
 
   task :run_installer do
-    installer_template = File.expand_path("../templates/installer.rb", __dir__)
```
Comment (rafaelfranca) [inline actiontext/lib/tasks/actiontext.rake]: Can we move this entire file to the generator?

File: actiontext/lib/rails/generators/actiontext/actiontext_generator.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+require "pathname"
+require "json"
+
+module Rails
+  module Generators
+    class ActiontextGenerator < Base
```
Comment (kaspth) [inline actiontext/lib/rails/generators/actiontext/actiontext_generator.rb]: `ActionText`? Or maybe this should even be `InstallGenerator`.

File: actiontext/lib/rails/generators/actiontext/actiontext_generator.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+      JS_PACKAGE = JSON.load(JS_PACKAGE_PATH)
+      JS_DEPENDENCIES = JS_PACKAGE["peerDependencies"].dup.merge \
+        JS_PACKAGE["name"] => "^#{JS_PACKAGE["version"]}"
+
+      source_root File.expand_path("templates", __dir__)
+
+      def install_javascript_dependencies
+        say "Installing JavaScript dependencies"
+        run "yarn add #{JS_DEPENDENCIES.map { |name, version| "#{name}@#{version}" }.join(" ")}",
+            abort_on_failure: true, capture: true
```
Comment (kaspth) [inline actiontext/lib/rails/generators/actiontext/actiontext_generator.rb]: Just indent by 2 instead of 4 spaces here.

**Comment by vinistock** [comment]
@kaspth @rafaelfranca: refactored the install generator to fit the commented requirements.

Please let me know if there are other improvements you'd like me to work on.

File: actiontext/lib/generators/action_text/install/install_generator.rb [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+                  "app/views/active_storage/blobs/_blob.html.erb"
+      end
+
+      def create_migrations
+        run "rake active_storage:install:migrations"
+        run "rake railties:install:migrations"
+        run "rake action_text:install:migrations"
+      end
+
+      def js_dependencies
```
Comment (kaspth) [inline actiontext/lib/generators/action_text/install/install_generator.rb]: This should be `private` otherwise Rails' generators will run the method as a generator method. Although that doesn't hurt, we should still be clear here.

File: actiontext/lib/generators/action_text/install/install_generator.rb [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        template "actiontext.scss", "app/assets/stylesheets/actiontext.scss"
+
+        copy_file "#{__dir__}/../../../../app/views/active_storage/blobs/_blob.html.erb",
+                  "app/views/active_storage/blobs/_blob.html.erb"
+      end
+
+      def create_migrations
+        run "rake active_storage:install:migrations"
+        run "rake railties:install:migrations"
+        run "rake action_text:install:migrations"
```
Comment (kaspth) [inline actiontext/lib/generators/action_text/install/install_generator.rb]: Should/can we assume that `action_text:install:migrations` has set up its dependencies properly and only run it?

File: actiontext/lib/generators/action_text/install/install_generator.rb [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+require "pathname"
+require "json"
+
+module ActionText
+  module Generators
+    class InstallGenerator < ::Rails::Generators::Base
+      source_root File.expand_path("templates", __dir__)
+
+      def install_javascript_dependencies
+        template "#{__dir__}/../../../../../railties/lib/rails/generators/rails/app/templates/bin/yarn.tt",
```
Comment (kaspth) [inline actiontext/lib/generators/action_text/install/install_generator.rb]: Don't think this will work once Action Text get's packaged as a gem. That file doesn't ship within it.

Maybe we can copy it when building the gem? What do you think @y-yagi?

File: actiontext/lib/generators/action_text/install/install_generator.rb [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+            Alternatively, install and setup the webpacker gem then rerun `bin/rails action_text:install`
+            to have these dependencies added automatically.
+          WARNING
+        end
+      end
+
+      def create_actiontext_files
+        template "actiontext.scss", "app/assets/stylesheets/actiontext.scss"
+
+        copy_file "#{__dir__}/../../../../app/views/active_storage/blobs/_blob.html.erb",
```
Comment (kaspth) [inline actiontext/lib/generators/action_text/install/install_generator.rb]: Let's extract something like `GEM_ROOT = "#{__dir__}/../../../.."` and use throughout. Or maybe `source_root` could work? There's got to be other generators that do this sort of thing that we can look to.

**Review by kaspth — COMMENTED** [review]
Cool, I like how this is coming together!

File: actiontext/lib/generators/action_text/install/install_generator.rb (hunk 7 above)
Comment (vinistock) [reply actiontext/lib/generators/action_text/install/install_generator.rb]: Which dependencies are you concerned about? I'd say if there are dependencies, it should check and warn the user if something is missing.

Comment (kaspth) [reply actiontext/lib/generators/action_text/install/install_generator.rb]: Sorry, I meant task dependencies. If `action_text:install:migrations` require `railties:install:migrations` and `active_storage:install:migrations` first, then shouldn't the Action Text task have those tasks as dependencies?

Comment (vinistock) [reply actiontext/lib/generators/action_text/install/install_generator.rb]: I have addressed two of your comments. However, I apologize, but it's still not clear to me what code changes implicate from having tasks as dependencies. I didn't find examples.

Comment (kaspth) [reply actiontext/lib/generators/action_text/install/install_generator.rb]: Let me rephrase just for the record. I would expect the Rake set up to be `task "action_text:install:migrations": [ "active_storage:install:migrations", "railties:install:migrations" ]`. In Rake's parlance when I run the Action Text migration task first run the Active Storage and Railties' migration tasks.

Anyway, I checked up on it and that's not how the tasks are defined. So this is fine.

File: actiontext/lib/generators/action_text/install/install_generator.rb [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+      def create_migrations
+        run "rake active_storage:install:migrations"
+        run "rake railties:install:migrations"
+        run "rake action_text:install:migrations"
+      end
+
+      hook_for :test_framework
+
+      private
+        GEM_ROOT = "#{__dir__}/../../../.."
```
Comment (kaspth) [inline actiontext/lib/generators/action_text/install/install_generator.rb]: Did you try something with `source_root` or similar instead of the constant? That would be more generator like.

File: actiontext/lib/generators/action_text/install/install_generator.rb [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+require "pathname"
+require "json"
+
+module ActionText
+  module Generators
+    class InstallGenerator < ::Rails::Generators::Base
+      source_root File.expand_path("templates", __dir__)
+
+      def install_javascript_dependencies
+        template "#{GEM_ROOT}/../railties/lib/rails/generators/rails/app/templates/bin/yarn.tt",
```
Comment (kaspth) [inline actiontext/lib/generators/action_text/install/install_generator.rb]: Again, I doubt this will work once Action Text is packaged as a gem. Try running `gem package` locally (or build or whatever the command is called). Then I think you can bundle the gem by pointing to it with `:path`. Finally try running `actiontext:install` from within your test app and see if it works.

Comment (vinistock) [reply actiontext/lib/generators/action_text/install/install_generator.rb]: How about changing the template copying to

```ruby
run "rake app:update:bin"
```

This will just create or update all bin files. Which would make sure `bin/yarn` exists. What do you think?

File: actiontext/lib/generators/action_text/install/install_generator.rb (hunk 10 above)
Comment (vinistock) [reply actiontext/lib/generators/action_text/install/install_generator.rb]: I looked at many generators and tried inspecting different methods such as `source_root`, `default_source_root`, `base_root but didn't find any of them that would point to the root of actiontext. They all either point to `__dir__` or to the railties directory.

File: actiontext/lib/generators/action_text/install/install_generator.rb (hunk 11 above)
Comment (kaspth) [reply actiontext/lib/generators/action_text/install/install_generator.rb]: I'm a little hesitant but don't know why. Let's go with it for now. @y-yagi if you can see anything wrong with this do follow up later 🙏

**Comment by kaspth** [comment]
@vinistock thanks for your patience and conquering the many reviews. Railties is one of the most complex and intricate bits to work on when contributing, so an extra big thank you for helping out here 🙏

**Comment by vinistock** [comment]
@kaspth I'm really happy to have helped! I'll keep an eye out for any feedback in case we need to polish it more. Thanks for all the collaboration!

**Comment by rafaelfranca** [comment]
Love this. Great work

File: actiontext/lib/generators/action_text/install/install_generator.rb [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+require "pathname"
+require "json"
+
+module ActionText
+  module Generators
+    class InstallGenerator < ::Rails::Generators::Base
+      source_root File.expand_path("templates", __dir__)
+
+      def install_javascript_dependencies
+        run "rake app:update:bin"
```
Comment (dhh) [inline actiontext/lib/generators/action_text/install/install_generator.rb:12]: When this is called running on master, it attempts to overwrite bin/rails and bin/rake with non-spring versions. Why do we need to call this at all? And if we do, how can we make it aware of spring?

Comment (abhaynikam) [reply actiontext/lib/generators/action_text/install/install_generator.rb:12]: I think it was added to add `bin/yarn`: https://github.com/rails/rails/pull/35085/commits/028d1611cdbc36b2650f80661cad6c79bb104051

Since Rails 6 already ships with all the necessary bin, I think we can skip it. I tested by creating new application from the master without updating binstub. Works fine. 

Also, on the master, ActionText installer won't work and silently fail because `@rails/actiontext@^6.1.0-alpha` isn't released yet and installer tries to install the latest version.(PR for same: #37824)

Comment (vinistock) [reply actiontext/lib/generators/action_text/install/install_generator.rb:12]: The addition of this evolved from an issue reported that the install of actiontext would fail silently without yarn installed. In my opinion, I do think it's a good idea to make sure the binstubs exist.

However, I'm not sure why the spring bits are not added. Could it be that `app:update:bin` is not matching what `rails new` generates for binstubs? And if so, is that a conscious decision or can we just update it?

Comment (abhaynikam) [reply actiontext/lib/generators/action_text/install/install_generator.rb:12]: @vinistock : I gave it a try by uninstalling yarn from my system on the master branch. 
Running `rails new` **without yarn**  fails with following message.
yarn-not-installed

Running `bundle exec rails action_text:install` **without yarn** also fails. The error is not user friendly but yet it doesn't silently fail.
fails-without-yarn-on-actiontext-install

I think we should go ahead and remove it and when such case or issue is reported we can rework on it to make it better. Here is the PR to remove it(I am holding that PR since morning, please do close it if not needed): https://github.com/rails/rails/pull/39233

Comment (vinistock) [reply actiontext/lib/generators/action_text/install/install_generator.rb:12]: I looked into this a bit further. The root of the issue is that we really only need to make sure `bin/yarn` exists, but we're running the `update:bin` generator which tries to override the binstubs generated by spring.

In my opinion, the ideal would be to surface a way of generating the yarn binstub exclusively (e.g.: `app:update:bin:yarn`). That way we wouldn't mess up the binstubs created by spring and yet be able to provide a nice friendly message about yarn not being installed.

That said, I have no objections to removing it until we figure it out appropriately.

Comment (dhh) [reply actiontext/lib/generators/action_text/install/install_generator.rb:12]: Let's have something that just ensures the yarn bin is available. Rather than overwriting all the other binstubs.

Comment (vinistock) [reply actiontext/lib/generators/action_text/install/install_generator.rb:12]: Sounds good. I'll put together a PR and we can iterate on it.
